packages/core/mesh/teleport-extension-object-sync/src/blob-store.ts

```diff
@@ -0,0 +1,155 @@
+
+    if (offset + length > metadata.length) {
+      throw new Error('Invalid range');
+    }
+
+    if (metadata.state === BlobMeta.State.FULLY_PRESENT) {
+      const file = this._getDataFile(id);
+      return file.read(offset, length);
+    } else if (options.offset === undefined && options.length === undefined) {
+      throw new Error('Blob not available');
+    }
+
+    const beginChunk = Math.floor(offset / metadata.chunkSize);
+    const endChunk = Math.ceil((offset + length) / metadata.chunkSize);
+
+    assert(metadata.bitfield, 'Bitfield not present');
+    assert(metadata.bitfield.length >= endChunk, 'Invalid bitfield length');
+
+    const present = BitField.count(metadata.bitfield, beginChunk, endChunk) === endChunk - beginChunk;
+
+    if (!present) {
+      throw new Error('Blob not available');
+    }
+
+    const file = this._getDataFile(id);
+    return file.read(offset, length);
+  }
+
+  async list(): Promise<BlobMeta[]> {
+    throw new Error('Not implemented');
+  }
+
+  async set(data: Uint8Array): Promise<BlobMeta> {
+    const id = new Uint8Array(await subtleCrypto.digest('SHA-256', data));
+    const bitFieldLength = Math.ceil(data.length / DEFAULT_CHUNK_SIZE / 8);
+    const bitfield = new Uint8Array(bitFieldLength).fill(0xff);
+
+    // Note: We need to calculate last byte of bitfield.
+    const amountOfChunksInLastByteOfBitfield = Math.ceil((data.length / DEFAULT_CHUNK_SIZE) % 8);
+    bitfield[bitFieldLength - 1] = 0xff << (8 - amountOfChunksInLastByteOfBitfield);
+
+    const meta: BlobMeta = {
+      id,
+      state: BlobMeta.State.FULLY_PRESENT,
+      length: data.length,
+      chunkSize: DEFAULT_CHUNK_SIZE,
+      bitfield,
+      created: new Date(),
+      updated: new Date(),
+    };
+
+    await this._getDataFile(id).write(0, Buffer.from(data));
+    await this._getMetaFile(id).write(0, Buffer.from(schema.getCodecForType('dxos.echo.blob.BlobMeta').encode(meta)));
+    return meta;
+  }
+
+  async setChunk(chunk: BlobChunk): Promise<BlobMeta> {
+    // Init metadata.
+    let meta = await this.getMeta(chunk.id);
+    if (!meta) {
+      assert(chunk.totalLength, 'totalLength is not present');
+      meta = {
+        id: chunk.id,
+        state: BlobMeta.State.PARTIALLY_PRESENT,
+        length: chunk.totalLength,
+        chunkSize: chunk.chunkSize ?? DEFAULT_CHUNK_SIZE,
+        created: new Date(),
+      };
+      meta.bitfield = new Uint8Array(Math.ceil(meta.length / meta.chunkSize / 8)).fill(0);
+    }
+
+    if (chunk.chunkSize && chunk.chunkSize !== meta.chunkSize) {
+      throw new Error('Invalid chunk size');
+    }
+
+    assert(meta.bitfield, 'Bitfield not present');
+    assert(chunk.chunkOffset !== undefined, 'chunkOffset is not present');
+
+    // Write chunk.
+    await this._getDataFile(chunk.id).write(chunk.chunkOffset, Buffer.from(chunk.payload));
+
+    // Update bitfield.
+    BitField.set(meta.bitfield, Math.floor(chunk.chunkOffset / meta.chunkSize), true);
```

Comment:
assert chunk offset is a multiple of chunk size

same for length so it aligns to chunks

Reply:
also if meta was previously set, assert that new chunkSize matches the old one

Reply:
> also if meta was previously set, assert that new chunkSize matches the old one

see line 121